### PR TITLE
Add support for "package extensions" to code loading

### PR DESCRIPTION
**Pull request by KristofferC**
This is an implementation of the proposal in https://github.com/JuliaLang/julia/issues/43119 (with some tweaks) that aims to replace Requires.jl for loading code based on the presence of other packages. It is a different implementation to https://github.com/JuliaLang/julia/pull/47040 and should work better in for example big environments where you expect to only load a small subset of packages.

In spirit, it is very similar to Requires.jl in that given a set of packages and a file, you load the file when those packages have been loaded in the julia session. The differences to Requires.jl are:

- The file with the conditional code gets precompiled in the same way as a normal package would. Requires.jl instead directly evaluates the code into the package.
- It is possible to use Pkg to set compatibility on the packages that "activates" the conditional code. This is a new feature in Pkg and could in theory also be used together with Requires.jl.
- It is "declarative" in the sense that all information requires id stored in the Project.toml and Manifest.toml file so it avoids having to run any arbitrary Julia code in e.g. `__init__` (which is done when using Requries.jl).

The added docs in code loading as well as the added docs in the accompanying Pkg PR (https://github.com/JuliaLang/Pkg.jl/pull/3264) can be used to get a more detail description of its usage.

A lot of the diff in this PR is due to the code loading being pretty awkward and would have been easier to be built on top of https://github.com/JuliaLang/julia/pull/46690 but I don't want to couple those PRs.

For reviewers: going through all the uuid lookups etc in the code loading is probably fairly uninteresting since it is mostly mechanical, the interesting parts are those around line 1080 - 1200 in `loading.jl` which deals with the insertion of gluepkgs "callbacks" and the loading of them.

An example of a real package moving from Requires.jl to using this system can be seen in https://github.com/KristofferC/PGFPlotsX.jl/compare/kc/glue. It should be pretty easy to be backwards compatible with Requires. You just put the `@require` inside a version check and include the file for the conditional file in the `@require` block.

One TODO that is left is that Pkg should precompile these "glue packages" during `Pkg.precompile` (which is often run automatically) in the case of them having a chance to be loaded in the current environment.

### CODE REVIEW

File: doc/src/manual/code-loading.md [hunk 1]
```diff
@@ -348,7 +348,46 @@ The subscripted `rootsᵢ`, `graphᵢ` and `pathsᵢ` variables correspond to th
...
+
+[gluepkgs]
+GlueFoo = "GlueDep"
+GlueBar = ["GlueDep", "OtherGlueDep"]
+...
+```
+
+The keys under `gluepkgs` are the name of the glue packages.
+They are loaded when all the packages on the right hand side (the glue dependencies) of the glue package are loaded.
+If a glue package only has one glue dependency the lit of glue dependencies can be written as just a string for breviety.
```
Comment (DilumAluthge) [inline doc/src/manual/code-loading.md:371]:
```suggestion
If a glue package only has one glue dependency, the list of glue dependencies can be written as just a string for brevity.
```

**Comment by KristofferC** [comment]
Pushed a bump to Pkg here to include https://github.com/JuliaLang/Pkg.jl/pull/3264 to run CI of these together.

**Comment by Seelengrab** [comment]
Does this have downsides/a tradeoff compared to #47040?

**Comment by KristofferC** [comment]
Some are discussed in https://github.com/JuliaLang/julia/pull/47040#issuecomment-1271305936

**Comment by Seelengrab** [comment]
Gotcha, I just thought there may be some other ones. Thinking about the description of this in the Pkg PR, what happens when there's glue needed between glue, e.g. when you have weak dependencies `A` and `B` and want different behavior for when only `A`, only `B` or both `A` and `B` are loaded? Would this need `GlueA`, `GlueB` and `GlueAB` modules..?

**Comment by KristofferC** [comment]
> when you have weak dependencies A and B and want different behavior for when only A, only B or both A and B are loaded? Would this need GlueA, GlueB and GlueAB modules..?

Yes.

Another question is, what if you want to activate extra functionality inside a glue package when some other package is loaded. Basically glue^2. The reason I think about this is that I saw this line:

https://github.com/JuliaPlots/Plots.jl/blob/7bb8f6aee1c93d4dd478cc0d596a804abec8a723/src/backends/pgfplotsx.jl#L1042-L1046

in Plots, where a `@requires` block is used inside another requires block. In theory, this could be specified like:

```toml
[gluedeps]
PGFPlotsX = "..."
Measurements = "..."

[gluepkgs]
PGFPlotsGlue = "PGFPlotsX"
PGFPlotsMeasurementGlue = ["PGFPlotsGlue", "Measurements"]
```

And `PGFPlotsMeasurementGlue` would load when both `PGFPlotsGlue` and `Measurements` are loaded.

**Comment by Seelengrab** [comment]
> Another question is, what if you want to activate extra functionality inside a glue package when some other package is loaded. Basically glue^2.

And that's exactly the reason why I'm more in favor of the other proposal :D No risk of accidentally going quadratic in terms of complexity, at the cost of penalizing larger environments a bit (which probably should just use a large sysimage anyway?).

**Comment by KristofferC** [comment]
@nanosoldier `runtests()`

**Comment by KristofferC** [comment]
> And that's exactly the reason why I'm more in favor of the other proposal :D No risk of accidentally going quadratic in terms of complexity, at the cost of penalizing larger environments a bit (which probably should just use a large sysimage anyway?).

The weak deps idea has some advantages and is more "powerful" since you have access to all information during precompile time, yes. But after thinking about this for quite some time I do think that the suggestion here is what should be tried first. Some reasons:

- It is conceptually similar to Requires which people seem happy to use (with the exception of its problems which are more side effects of the implementation). This also makes it quite easy to be backward compatible with Requires (which the Pkg PR has docs for how to be).
- When the "conditional features" (or glue packages) are separate modules that are loaded, the extra stuff is more or less "pure" additions to what the base package offers. I think that would mean that it would be easier to test each thing separately.
- I think the amount of recompilation that would be required with the weak deps thing is pretty bad. If you remove a package, you might have to recompile your whole environment. With glue deps, adding a package will at most cause you to compile new glue packages (which are small), and removing a package will never require new compilation.
- I think this will on average give users a smaller load time. Weak dependencies really punish you if you don't have super clean environment usage and a lot of people just don't use Julia like that. We can say "oh but people should make sysimages or  have smaller envs etc" but if that is just not how the majority of people use Julia, that should be taken into account.'

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/47695#issuecomment-1328239201) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/f53ca51_vs_726bbd7/report.html).

**Comment by KristofferC** [comment]
PkgEval looks good.

File: doc/src/manual/code-loading.md [hunk 2]
```diff
@@ -348,7 +348,47 @@ The subscripted `rootsᵢ`, `graphᵢ` and `pathsᵢ` variables correspond to th
 2. Packages in non-primary environments can end up using incompatible versions of their dependencies even if their own environments are entirely compatible. This can happen when one of their dependencies is shadowed by a version in an earlier environment in the stack (either by graph or path, or both).
 
 Since the primary environment is typically the environment of a project you're working on, while environments later in the stack contain additional tools, this is the right trade-off: it's better to break your development tools but keep the project working. When such incompatibilities occur, you'll typically want to upgrade your dev tools to versions that are compatible with the main project.
+### "Glue" packages and dependencies
```
Comment (IanButterworth) [inline doc/src/manual/code-loading.md:351]: Obviously there are more changes to like this, but just to mark it as I think this is where the discussion settled
```suggestion
### "Weak" dependencies and  "Glue" packages
```

File: doc/src/manual/code-loading.md [hunk 3]
```diff
@@ -348,7 +348,47 @@ The subscripted `rootsᵢ`, `graphᵢ` and `pathsᵢ` variables correspond to th
...
 
+A "glue package" is a module that is automatically loaded when a specified set of other packages (called "glue dependencies") are loaded in the current Julia session.
+These are defined by adding the following two sections to a package's `Project.toml` file:
+
+```toml
+name = "MyPackage"
+
+[gluedeps]
+GlueDep = "c9a23..." # uuid
+OtherGlueDep = "862e..." # uuid
```
Comment (IanButterworth) [inline doc/src/manual/code-loading.md:361]:
```suggestion
[weakdeps]
WeakDep = "c9a23..." # uuid
OtherWeakDep = "862e..." # uuid
```

File: doc/src/manual/code-loading.md [hunk 4]
```diff
@@ -348,7 +348,47 @@ The subscripted `rootsᵢ`, `graphᵢ` and `pathsᵢ` variables correspond to th
...
+
+[gluepkgs]
+GlueFoo = "GlueDep"
+GlueBar = ["GlueDep", "OtherGlueDep"]
```
Comment (IanButterworth) [inline doc/src/manual/code-loading.md:365]: I think setting up a convention of *Glue would be better here. i.e. `FooGlue` reads like _the glue for Foo_
```suggestion
WeakDepGlue = "WeakDep"
FooGlue = ["WeakDep", "OtherWeakDep"]
```

File: doc/src/manual/code-loading.md [hunk 5]
```diff
@@ -348,7 +348,47 @@ The subscripted `rootsᵢ`, `graphᵢ` and `pathsᵢ` variables correspond to th
...
+...
+```
+
+The keys under `gluepkgs` are the name of the glue packages.
+They are loaded when all the packages on the right hand side (the glue dependencies) of the glue package are loaded.
```
Comment (IanButterworth) [inline doc/src/manual/code-loading.md:370]:
```suggestion
They are loaded when all the packages on the right hand side (the weak dependencies) of the glue package are loaded.
```

**Comment by KristofferC** [comment]
For those following along, it is likely that "glue packages" will be renamed to "package extensions".

File: NEWS.md [hunk 6]
```diff
@@ -4,7 +4,11 @@ Julia v1.10 Release Notes
 New language features
 ---------------------
 
-
+New feature for packages that allows loading a piece of code based on other
+packages being loaded in the Julia session.
```
[the remark this answers is not in the document]
Comment (KristofferC) [reply NEWS.md]: "based on other packages are loaded"?

Sounds weird to me.

[the remark this answers is not in the document]
Comment (IanButterworth) [reply NEWS.md]: Ah thought there was a "whether" other packages. Your way is better

**Comment by KristofferC** [comment]
I feel this is getting pretty ready. Happy for some not eyes on it though. The steps for "completion" are:

- Drop the Pkg bump here.
- Merge the Pkg PR when nighties have updated.
- Bump Pkg on master.
- Update Registrator with https://github.com/JuliaRegistries/RegistryTools.jl/pull/74 (might require https://github.com/JuliaRegistries/RegistryTools.jl/issues/75 if we don't want to run Registrator on julia nightly).
- Profit

**Review by IanButterworth — APPROVED** [review]
Awesome.

File: doc/src/manual/code-loading.md [hunk 7]
```diff
@@ -348,7 +348,46 @@ The subscripted `rootsᵢ`, `graphᵢ` and `pathsᵢ` variables correspond to th
 2. Packages in non-primary environments can end up using incompatible versions of their dependencies even if their own environments are entirely compatible. This can happen when one of their dependencies is shadowed by a version in an earlier environment in the stack (either by graph or path, or both).
 
 Since the primary environment is typically the environment of a project you're working on, while environments later in the stack contain additional tools, this is the right trade-off: it's better to break your development tools but keep the project working. When such incompatibilities occur, you'll typically want to upgrade your dev tools to versions that are compatible with the main project.
+### "Extension"s
 
+An "extension" is a module that is automatically loaded when a specified set of other packages (its "extension dependencies") are loaded in the current Julia session. The extension dependencies of an extension is a subset of those packages listed under the `[weakdeps]` section of a Project file. Extensions are defined under the `[extensions]` section in the project file:
```
Comment (fingolfin) [inline doc/src/manual/code-loading.md:353]:
```suggestion
An "extension" is a module that is automatically loaded when a specified set of other packages (its "extension dependencies") are loaded in the current Julia session. The extension dependencies of an extension are a subset of those packages listed under the `[weakdeps]` section of a Project file. Extensions are defined under the `[extensions]` section in the project file:
```

File: base/loading.jl [hunk 8]
```diff
@@ -972,6 +1016,154 @@ function run_package_callbacks(modkey::PkgId)
...
+                        end
+                    end
+                end
+                if ext_not_allowed_load !== nothing
+                    @debug "Extension $(extid.id.name) of $(extid.parentid.name) not loaded due to \
+                            $(ext_not_allowed_load.name) loaded in environment lower in load path"
+                else
+                    require(extid.id)
+                    @debug "Extension $(extid.id.name) of $(extid.parentid.name) loaded"
+                end
```
Comment (johnnychen94) [inline base/loading.jl:1131]: Not sure if this is a deliberate design so wanna ask it here. Does it sound like a good idea to make `Plotting.ContourExt` a "normal" submodule? And if not loaded, `Plotting.ContourExt` keeps to be `nothing`.

**Pull request merged**